### PR TITLE
Support schema cache when querying all measurement of devices using template

### PR DESCRIPTION
**Pull request by MarcosZyk**
## Description

Currently, a device using schema template may be cached in DataNodeSchemaCache, but the cache is only used for data insertion.

Enable data query to benefit from such device cache when querying all measurements of the device, e.g. select * from root.db.d, with schema template t1 set on db.

### CODE REVIEW

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java [hunk 1]
```diff
@@ -107,13 +107,25 @@ public PartialPath(String path, boolean needSplit) {
...
 
+  // e.g. root.db.d.s, root.db.d.*, root.db.d.s*, not include patterns like root.db.d.**
+  public boolean hasExplicitDevice() {
+    for (int i = 0; i < nodes.length - 1; i++) {
+      // *, ** , d*, *d*
+      if (PathPatternUtil.hasWildcard(nodes[i])) {
+        return false;
+      }
+    }
+    return !nodes[nodes.length - 1].equals(MULTI_LEVEL_PATH_WILDCARD);
```
Comment (Cpaulyz) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java:126]:
```suggestion
    if(nodes[nodes.length - 1].equals(MULTI_LEVEL_PATH_WILDCARD)){
        return false;
    }
    for (int i = 0; i < nodes.length - 1; i++) {
      // *, ** , d*, *d*
      if (PathPatternUtil.hasWildcard(nodes[i])) {
        return false;
      }
    }
    return true;
```

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternUtil.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+public class PathPatternUtil {
+
+  private PathPatternUtil() {}
+
+  /**
+   * The input string is a single node of a path pattern. Return true if the node may be a
+   * patternNode that can match batch explicit node names. e.g. *, e.g. *, **, d*, *d*.
+   */
+  public static boolean hasWildcard(String node) {
+    return node.startsWith(ONE_LEVEL_PATH_WILDCARD) || node.endsWith(ONE_LEVEL_PATH_WILDCARD);
```
Comment (Cpaulyz) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternUtil.java:36]:
```suggestion
    return node.contains(ONE_LEVEL_PATH_WILDCARD);
```

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternUtil.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  /** The input patternNode shall be a string starts or ends with *, e.g. *, **, d*, *d* */
+  public static boolean isNodeMatch(String patternNode, String nodeName) {
```
Comment (Cpaulyz) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternUtil.java:40]:
```suggestion
  /** 
  * Determine if a wildcard matches a node name.
  *
  * @param wildcardPattern must be a string starts or ends with *, e.g. *, **, d*, *d* 
  * @param nodeName node to match
  */
  public static boolean isWildcardMatch(String wildcardPattern, String nodeName) {
```

**Review by Cpaulyz — COMMENTED** [review]
Some code smells and execution optimization. PTAL~

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java (hunk 1 above)
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java]: Fixed.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternUtil.java (hunk 2 above)
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternUtil.java:36]: The new judgement will be kept and all other related logic has been modified and rely on this method.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternUtil.java (hunk 3 above)
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternUtil.java]: Fixed.

**Review by Cpaulyz — APPROVED** [review]
LGTM

**Pull request merged**